boskos/boskos.go

```diff
@@ -0,0 +1,302 @@
+			fmt.Fprint(res, string(projJSON))
+		} else {
+			logrus.Infof("No available project")
+			res.WriteHeader(http.StatusNoContent)
+			fmt.Fprint(res, "No available project, please try again later\n")
+		}
+	}
+}
+
+func handleReturnProject(myRanch *RanchoSantaFe) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleReturnProject").Infof("From %v", req.RemoteAddr)
+
+		// validate vars
+		job := req.URL.Query().Get("job")
+		if job == "" {
+			http.Error(res, "job must present in the request", http.StatusBadRequest)
+			return
+		}
+
+		logrus.Infof("Job %v returns a project", job)
+		if err := myRanch.ReturnCattle(job); err == nil {
+			fmt.Fprint(res, "Successful\n")
+		} else {
+			logrus.WithError(err).Info("Return failed")
+			http.Error(res, error.Error(err), http.StatusInternalServerError)
+		}
+	}
+}
+
+func handleHeartbeat(myRanch *RanchoSantaFe) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleHeartbeat").Infof("From %v", req.RemoteAddr)
+		if !myRanch.CheckLease() {
+			res.WriteHeader(http.StatusNoContent)
+		}
+	}
+}
+
+type RanchoSantaFe struct {
+	MyCattles    []*GCPProject
+	JobToCattles map[string]*GCPProject
+	CattleLock   sync.RWMutex
+	Storage      string
+}
+
+func NewProject(projName string) *GCPProject {
+	newProject := &GCPProject{
+		Name: projName,
+	}
+	return newProject
+}
+
+func NewRancho(config string, storage string) *RanchoSantaFe {
+	file, err := os.Open(config)
+	if err != nil {
+		logrus.WithError(err).Fatal(fmt.Sprintf("Fail to open config file %v.", config))
+	}
+	defer file.Close()
+
+	newRanch := &RanchoSantaFe{
+		Storage:      storage,
+		JobToCattles: make(map[string]*GCPProject),
+	}
+
+	if storage != "" {
+		buf, err := ioutil.ReadFile(storage)
+		if err == nil {
+			logrus.Infof("Current state: %v.", buf)
+			err = json.Unmarshal(buf, newRanch)
+			if err != nil {
+				logrus.WithError(err).Fatal(fmt.Sprintf("Fail to unmarshal state from %v.", storage))
+			}
+		} else if !os.IsNotExist(err) {
+			logrus.WithError(err).Fatal(fmt.Sprintf("Fail to read state from %v.", storage))
+		}
+	}
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		proj := scanner.Text()
+		for _, exist := range newRanch.JobToCattles {
+			if proj == exist.Name {
+				continue
+			}
+		}
+
+		// append the free ones
+		newRanch.MyCattles = append(newRanch.MyCattles, NewProject(proj))
+	}
+
+	return newRanch
+}
+
+func (myRanch *RanchoSantaFe) LogStatus() {
+	myRanch.CattleLock.RLock()
+	defer myRanch.CattleLock.RUnlock()
+
+	for _, proj := range myRanch.MyCattles {
+		projJSON, _ := json.Marshal(proj)
+		logrus.Infof("Current Projects : %v", string(projJSON))
+	}
+
+	for job, proj := range myRanch.JobToCattles {
+		projJSON, _ := json.Marshal(proj)
+		logrus.Infof("Current Map : %v - %v", job, string(projJSON))
+	}
+}
+
+func (myRanch *RanchoSantaFe) SaveState() {
```

Comment:
it's called within lock scope already, for all the instances. Is go mutex recursive?